File: di-ipv-cimit-stub/lambdas/stub-management/src/main/java/uk/gov/di/ipv/core/stubmanagement/service/UserService.java
package uk.gov.di.ipv.core.stubmanagement.service;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import software.amazon.awssdk.utils.StringUtils;
import uk.gov.di.ipv.core.library.model.UserCisRequest;
import uk.gov.di.ipv.core.library.persistence.items.CimitStubItem;
import uk.gov.di.ipv.core.library.service.CimitStubItemService;
import uk.gov.di.ipv.core.library.service.ConfigService;
import uk.gov.di.ipv.core.stubmanagement.exceptions.BadRequestException;

import java.util.List;

public class UserService {

    private static final Logger LOGGER = LogManager.getLogger();

    private final ConfigService configService;
    private final CimitStubItemService cimitStubItemService;

    public UserService() {
        this.configService = new ConfigService();
        this.cimitStubItemService = new CimitStubItemService(configService);
    }

    public UserService(ConfigService configService, CimitStubItemService cimitStubItemService) {
        this.configService = configService;
        this.cimitStubItemService = cimitStubItemService;
    }

    public void addUserCis(String userId, List<UserCisRequest> userCisRequest) {
        checkCICodes(userCisRequest);
        userCisRequest.forEach(
                ciRequest ->
                        cimitStubItemService.persistCimitStubItem(
                                CimitStubItem.fromUserCiRequest(ciRequest, userId)));
        LOGGER.info("Inserted User CI data to the Cimit Stub DynamoDB Table.");
    }

    public void updateUserCis(String userId, List<UserCisRequest> userCisRequest) {
        checkCICodes(userCisRequest);
        List<CimitStubItem> cimitStubItems = cimitStubItemService.getCIsForUserId(userId);
        if (!cimitStubItems.isEmpty()) {
            deleteCimitStubItems(cimitStubItems);
        }
        userCisRequest.forEach(
                ciRequest ->
                        cimitStubItemService.persistCimitStubItem(
                                CimitStubItem.fromUserCiRequest(ciRequest, userId)));
    }

    private void deleteCimitStubItems(List<CimitStubItem> cimitStubItems) {
        cimitStubItems.forEach(
                item ->
                        cimitStubItemService.deleteCimitStubItem(
                                item.getUserId(), item.getContraIndicatorCode()));
    }

    private static void checkCICodes(List<UserCisRequest> userCisRequest) {
        if (userCisRequest.stream().anyMatch(user -> StringUtils.isEmpty(user.getCode()))) {
            throw new BadRequestException("CI codes cannot be empty.");
        }
    }
}

File: di-ipv-cimit-stub-ts/lambdas/src/external-api/stub-management/service/userService.ts
import { CimitStubItem } from "../../../common/contraIndicatorTypes";
import * as cimitStubItemService from "../../../common/cimitStubItemService";

export interface UserCisRequest {
  code: string;
  issuanceDate: string;
  issuer: string;
  mitigations: string[];
  document: string;
  txn: string;
}

export async function addUserCis(
  userId: string,
  userCisRequests: UserCisRequest[],
): Promise<void> {
  for (const ciRequest of userCisRequests) {
    const cimitStubItem = cimitStubItemService.fromUserCisRequest(
      ciRequest,
      userId,
    );
    await cimitStubItemService.persistCimitStubItem(cimitStubItem);
  }

  console.log("Inserted User CI data to the Cimit Stub DynamoDB Table.");
}

export async function updateUserCis(
  userId: string,
  userCisRequests: UserCisRequest[],
): Promise<void> {
  const cimitStubItems = await cimitStubItemService.getCIsForUserId(userId);

  if (cimitStubItems.length > 0) {
    await deleteCimitStubItems(cimitStubItems);
  }

  for (const ciRequest of userCisRequests) {
    const cimitStubItem = cimitStubItemService.fromUserCisRequest(
      ciRequest,
      userId,
    );
    await cimitStubItemService.persistCimitStubItem(cimitStubItem);
  }
}

async function deleteCimitStubItems(
  cimitStubItems: CimitStubItem[],
): Promise<void> {
  for (const item of cimitStubItems) {
    await cimitStubItemService.deleteCimitStubItem(item.userId, item.sortKey);
  }
}
